Improve error diagnostics of X509V3 conf parsing, CMP, and other apps

DDvO opened the pull request:
This fixes #12268
and various other small issues I recently found using the CMP CLI and other apps:

* Error messages on parsing X.509v3 extension specs in config files where often weird. For instance for
```
extendedKeyUsage = clientAuth, serverAuth, wrong
```
the diagnostics was:
```
asn1 encoding routines:crypto/asn1/a_object.c:73:CMP error: first num too large
X509 V3 routines:crypto/x509/v3_extku.c:98:CMP error: invalid object identifier : section:<NULL>,name:unknown_key_usage_name,value:<NULL>
X509 V3 routines:crypto/x509/v3_conf.c:50:CMP error: error in extension : name=extendedKeyUsage, value=clientAuth, serverAuth, unknown_key_usage_name
```
With the fixes done here, this becomes:
```
object identifier routines:crypto/objects/obj_dat.c:376:CMP error: unknown object name
X509 V3 routines:crypto/x509/v3_extku.c:98:CMP error: invalid object identifier : unknown_key_usage_name
X509 V3 routines:crypto/x509/v3_conf.c:50:CMP error: error in extension : section=myKeyUsages, name=extendedKeyUsage, value=clientAuth, serverAuth, unknown_key_usage_name
```

* Correct and extend the diagnostics of `parse_name()` in `apps/lib/apps.c`
* Defer the so far premature diagnostic output of `apps/cmp.c` on server+proxy to be contacted.

paulidale commented on apps/lib/apps.c:
```diff
@@ -1644,8 +1648,8 @@ X509_NAME *parse_name(const char *cp, long chtype, int canmulti)
             BIO_printf(bio_err,
-                       "%s: Hit end of string before finding the '='\n",
-                       opt_getprog());
+                       "%s: Hit end of %d name string before finding the '='\n",
```
There ought to be a better word that "Hit".  What about "Reached"?

DDvO replied:
Yes, in this case we can simply (and more concisely) replace "Hit end of" by "End of" - done that.

paulidale commented on crypto/objects/obj_dat.c:
```diff
@@ -372,6 +372,10 @@ ASN1_OBJECT *OBJ_txt2obj(const char *s, int no_name)
         if (((nid = OBJ_sn2nid(s)) != NID_undef) ||
             ((nid = OBJ_ln2nid(s)) != NID_undef))
             return OBJ_nid2obj(nid);
+        if (*s < '0' || *s > '9') {
```
Might `if (!ossl_isdigit(*s))` be safer since it is character set aware?

DDvO replied:
Good point - done.

paulidale commented on crypto/x509/v3_cpols.c:
```diff
@@ -116,11 +117,10 @@ static STACK_OF(POLICYINFO) *r2i_certpol(X509V3_EXT_METHOD *method,
             X509V3err(X509V3_F_R2I_CERTPOL,
                       X509V3_R_INVALID_POLICY_IDENTIFIER);
-            X509V3_conf_err(cnf);
+            X509V3_conf_err2(cnf);
```
Is there a better name than `X509V3_conf_err2`?

DDvO replied:
How about `X509v3_conf_add_error_name_value`?

DDvO replied:
I just did this tentatively.

paulidale commented on include/openssl/x509v3.h:
```diff
@@ -331,10 +331,6 @@ struct ISSUING_DIST_POINT_st {
 /* onlysomereasons present */
 # define IDP_REASONS     0x40
 
-# define X509V3_conf_err(val) ERR_add_error_data(6, \
```
This is a breaking API change.

DDvO replied:
Well, that macro is exported but undocumented.
Shall I move its definition back to `include/openssl/x509v3.h`?

DDvO replied:
Just did this tentatively as well.

levitte commented on apps/lib/apps.c:
```diff
@@ -1648,7 +1648,7 @@ X509_NAME *parse_name(const char *cp, int chtype, int canmulti,
         if (*cp == '\0') {
             BIO_printf(bio_err,
-                       "%s: Hit end of %d name string before finding the '='\n",
+                       "%s: End of %d name string before finding the '='\n",
```
`"%s: %d name string doesn't contain a '='\n"`

DDvO replied:
> `"%s: %d name string doesn't contain a '='\n"`

This is not general enough because there may be occurrences of `=` (for other RDNs) before."
Just improved this to:
```
            BIO_printf(bio_err,
                       "%s: Missing '=' after RDN type string '%s' in %s name string\n",
                       opt_getprog(), typestr, desc);

```
(which required terminating the `typestr` just before that).

levitte replied:
Even better!

DDvO commented:
Thanks @paulidale for reviewing this.
Dues this look okay now?

DDvO commented:
For whatever reason, the Travis builds have been queued, for whatever strange reason, for two days.
I've just rebased, squashed, and force-pushed on this PR; hope Travis will do better this time.

DDvO commented:
@paulidale, can you still have a quick look at the changes you requested and my reaction and approve that before you leave?

In any case, have a nice and restful vacation!

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

DDvO commented:
Merged - thanks @paulidale
